qcodes/instrument/abstract.py

```diff
@@ -0,0 +1,101 @@
+    """
+    Errors with abstract parameter utilization should
+    raise this exception class.
+    """
+
+
+def abstract_instrument(cls: Type) -> Type:
+    """
+    A class decorator to create an abstract instrument. Abstract
+    instruments are allowed to have abstract parameters, but
+    their subclasses do not. This works by replacing the
+    'add_parameter' and '__init_subclass__' methods of the class
+    dynamically.
+
+    Args:
+        cls: The class to be decorated
+
+    Returns:
+        The decorated class
+    """
+
+    def __init_subclass__(sub_cls: Type) -> None:
+
+        __init__ = sub_cls.__init__
+
+        def __init_new__(self, *args: Any, **kwargs: Any) -> None:
+            """
+            Subclasses of an abstract instrument should check
+            after initialization whether there still are
+            abstract parameters. If there are, we should raise
+            an exception.
+            """
+            __init__(self, *args, **kwargs)
+            # after the usual initialization...
+
+            abstract_parameters = [
+                parameter.name for parameter in self.parameters.values()
+                if isinstance(parameter, AbstractParameter)
+            ]
+
+            if any(abstract_parameters):
+                cls_name = sub_cls.__name__
+
+                raise AbstractParameterException(
+                    f"Class '{cls_name}' has un-implemented Abstract Parameter(s): " +
+                    ", ".join([f"'{name}'" for name in abstract_parameters])
+                )
+
+        sub_cls.__init__ = __init_new__
+
+    original_add_parameter = cls.add_parameter
+
+    def add_parameter(
+            self, name: str, parameter_class: type = Parameter,
+            **kwargs: Any
+    ) -> None:
+
+        existing_parameter = self.parameters.get(name, None)
+
+        if isinstance(existing_parameter, AbstractParameter):
+            # For abstract parameters, we define special behavior.
+            existing_unit = getattr(existing_parameter, "unit", None)
+            new_unit = kwargs.get("unit", None)
+
+            if existing_unit and existing_unit != new_unit:
+                raise AbstractParameterException(
+                    f"The unit of the parameter '{name}' is '{new_unit}', "
+                    f"which is inconsistent with the unit '{existing_unit}' "
+                    f"specified earlier. This is usually because a driver "
+                    f"is a subclass of a baseclass which defines a parameter "
+                    f"of the same name but with different units"
+                )
+
+            param = parameter_class(name=name, instrument=self, **kwargs)
+            self.parameters[name] = param
+
+        else:
+            # If it is a parameter other then abstract parameter, call the original
+            # method
+            original_add_parameter(
+                self, name, parameter_class, **kwargs
+            )
```

Comment:
shall we call the original method anyway, also for the abstract parameter case?
```suggestion
            del self.parameters[name]

        original_add_parameter(
            self, name, parameter_class, **kwargs
        )
```

Reply:
We can, but not as-is. The original 'add_parameter' will complain that there already is a parameter with the name you are adding. The only way this will work is if in the decorated case we remove the abstract parameter first. It will work either way so I am fine with this solution.

Reply:
> The only way this will work is if in the decorated case we remove the abstract parameter first

that's what i suggest in the green snippet indeed :)